modyn/evaluator/internal/metric_factory.py

```diff
@@ -1,23 +1,45 @@
-from modyn.config.schema.pipeline import validate_metric_config_json
-from modyn.evaluator.internal.metrics import AbstractEvaluationMetric, AbstractHolisticMetric, Accuracy, F1Score, RocAuc
+from typing import overload
+
+from modyn.config.schema.pipeline.evaluation.metrics import (
+    MetricConfig,
+    validate_metric_config_json,
+)
+from modyn.evaluator.internal.metrics import (
+    AbstractEvaluationMetric,
+    AbstractHolisticMetric,
+    Accuracy,
+    F1Score,
+    RocAuc,
+)
 
 all_metrics: set[type[AbstractEvaluationMetric]] = {Accuracy, F1Score, RocAuc}
 
 
 class MetricFactory:
+    @overload
+    @staticmethod
+    def get_evaluation_metric(config: str) -> AbstractEvaluationMetric: ...
+
+    @overload
     @staticmethod
-    def get_evaluation_metric(config_json: str) -> AbstractEvaluationMetric:
-        parsed_metric_config = validate_metric_config_json(config_json)
+    def get_evaluation_metric(config: MetricConfig) -> AbstractEvaluationMetric: ...
+
+    @staticmethod
+    def get_evaluation_metric(config: str | MetricConfig) -> AbstractEvaluationMetric:
+        if isinstance(config, str):
+            parsed_metric_config = validate_metric_config_json(config)
+        else:
+            parsed_metric_config = config
+
         for metric in all_metrics:
             if metric.__name__.lower() == parsed_metric_config.name.lower():
                 return metric(parsed_metric_config)
         raise NotImplementedError(f"Metric {parsed_metric_config.name} is not available!")
 
     @staticmethod
-    def prepare_metrics(metrics: list[AbstractEvaluationMetric]) -> bool:
-        contains_holistic = False
-        for metric in metrics:
+    def prepare_metrics(metrics: dict[str, AbstractEvaluationMetric]) -> bool:
+        for metric in metrics.values():
             metric.deserialize_evaluation_transformer()
             if isinstance(metric, AbstractHolisticMetric):
                 return True
-        return contains_holistic
+        return False
```

Comment:
This will return true as soon as we reach a holistic metric. This means for metrics after that we won't call the deserialize function. I think we should instead change return True to contains_holistic = True and return that at the end

Reply:
good catch!